src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,329 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "snippets/lowered/pass/extract_loop_invariants.hpp"
+#include "snippets/utils.hpp"
+#include "snippets/lowered/linear_ir.hpp"
+#include "snippets/lowered/loop_manager.hpp"
+#include "snippets/snippets_isa.hpp"
+#include "snippets/utils.hpp"
+#include "snippets/itt.hpp"
+
+namespace ov {
+namespace snippets {
+namespace lowered {
+namespace pass {
+
+bool ExtractLoopInvariants::run(LinearIR& linear_ir) {
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::ExtractLoopInvariants")
+    bool modified = false;
+    const auto& loop_manager = linear_ir.get_loop_manager();
+
+    auto remove_last_loop_id = [](const std::shared_ptr<Expression>& expr) {
+        auto loop_ids = expr->get_loop_ids();
+        loop_ids.pop_back();
+        expr->set_loop_ids(loop_ids);
+    };
+
+    // move invariant expr to top(outside) of current loop
+    for (auto expr_it = linear_ir.begin(); expr_it != linear_ir.end(); expr_it++) {
+        const auto& expr = *expr_it;
+
+        // if there is scalar need to be moved(which has 0 input), move scalar with its consumer, as scalar is before consumer due to MoveScalarToConsumer
+        size_t input_num = expr->get_input_count();
+        if (input_num <= 0) {
+            continue;
+        }
+        const auto& current_expr_loops = expr->get_loop_ids();
+        const auto current_loop_depth = current_expr_loops.size();
+        // move from inner loop to outer loop
+        if (current_loop_depth != 2) {
```

Comment:
I see very complex algorithm here. Let's think if we can make it easier and more generic (I see the limitations for really `loop_depth == 2)😃 

Since it's task about loops, what's about work with `LoopInfo` here?

We should iterate by `LinearIR` and extract expression from innermost Loop to outermost while it's possible.
For example, we can make a pause when `next_expr->get_loop_ids().size() < current_expr->get_loop_ids().size()` - it means that next expression has less `LoopIDs` and it is Loop outside -> current_expr is last expression of the current innermost Loop with ID  `current_expr->get_loop_ids().back()`.

Take `LoopInfo` from `LoopManager` by this `ID`.

Using `get_entry_points` - we take all entry points of the current Loop. They're are `netry` points of the Loop. And only they can be extract firstly. Let's started from them.

Take one `entry_point` (`LoopPort`): if all other input ports of the expressions are `LoopPorts` too, we can continue. Check if all input [strides](https://github.com/openvinotoolkit/openvino/blob/ac27cd6fe75f18373381e2fb94b826889c1fd26d/src/common/snippets/src/lowered/pass/init_loops.cpp#L75) is equal to 1, we can extract `Expression` from the current Loop: you need just:
1. Update `LoopPorts` of the `LoopInfo` using [update_loop_ports](https://github.com/openvinotoolkit/openvino/blob/4123b6b9dea312a1e75e253d4fb2e09d8a133b4a/src/common/snippets/include/snippets/lowered/loop_manager.hpp#L184) (`actutal_port` is input port of the current expr, `target_ports` is set of input ports of consumers of this expressions)
2. Move this expre before iterator of pseudo `LoopBegin` (`loop_info->get_loop_bounds` - we will find iterators of pseudo `LoopBegin` and `LoopEnd` (I mean `pseudo`. because we don't these expresssions in LinearIR for now and at the moment `LoopBegin` iterator is iterator of the first expression in the loop and `LoopEnd` - is next iterator of the latest)).
3. Change `loop_ids` of the expression (remove last ID from `get_loop_ids()`)
4. Completed!

After analysis of `entry_points` we can check `exit_points` of `LoopInfo`: if the parent is out of loop (was extracted a few moments ago), we can extract this node as well (I'm not sure that it's correct in general, could you please verify?)

Also need to check: if there is no expressions in this Loop: remove the Loop from LoopManager.

It's general idea. Probably you already thought about that but faced with some problems. If it's so, could you please share with us? 😃 Because at first glance it is not obvious.

Thus, you don't need to introduce `delete_loop_ports` and `insert_loop_ports` because when you extract loop ports, you always inserted new ports instead of the previous - is mechanism that implemented by  `update_loop_ports`.

Reply:
@a-sidorova Done as discussed offline. Extract each loop, from inner to outer, which make it generic that extract expr from multiple loop depth. New interface are added to update inner loop and outer loop info. Thanks a lot Alexandra!